File: app/src/main/java/com/icostel/arhitecturesample/ui/listusers/UserAdapter.java
package com.icostel.arhitecturesample.ui.listusers;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bumptech.glide.GenericTransitionOptions;
import com.icostel.arhitecturesample.R;
import com.icostel.arhitecturesample.di.modules.GlideApp;
import com.icostel.arhitecturesample.utils.AnimationFactory;
import com.icostel.arhitecturesample.utils.ImageRequestListener;
import com.icostel.arhitecturesample.utils.livedata.SingleLiveEvent;
import com.icostel.arhitecturesample.view.model.User;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

public class UserAdapter extends RecyclerView.Adapter<UserAdapter.UserViewHolder> {

    private static final String TAG = UserAdapter.class.getCanonicalName();

    private final List<User> users = new ArrayList<>();
    private Context context;
    private SingleLiveEvent<User> selectedUserLive = new SingleLiveEvent<>();

    UserAdapter(Context context) {
        this.context = context;
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View charityItemView = LayoutInflater.from(this.context).inflate(R.layout.item_user, parent, false);
        return new UserViewHolder(charityItemView);
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        holder.bindUserViewHolder(users.get(position));
    }

    void updateUserList(List<User> newUserList) {
        Timber.d("updateUserList(), size: %d", newUserList.size());
        final UserDiffCallback diffCallback = new UserDiffCallback(users, newUserList);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(diffCallback);
        users.clear();
        users.addAll(newUserList);
        notifyDataSetChanged();
        // push only the new items after the diff is done
        diffResult.dispatchUpdatesTo(this);
    }

    SingleLiveEvent<User> getSelectedUserLive() {
        return selectedUserLive;
    }

    @Override
    public int getItemCount() {
        return users.size();
    }

    class UserViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.root_view)
        View rootView;

        @BindView(R.id.first_name)
        TextView firstName;

        @BindView(R.id.user_image)
        AppCompatImageView userImage;

        @BindView(R.id.age)
        TextView age;

        UserViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        private ActivityOptions getTransitionOptions() {
            return ActivityOptions.makeSceneTransitionAnimation((Activity) itemView.getContext(), userImage, "user_image_transition");
        }

        void bindUserViewHolder(User user) {
            firstName.setText(user.getFirstName());
            age.setText(context.getString(R.string.age, String.valueOf(user.getAge())));
            GlideApp.with(context)
                    .load(user.getResourceUrl())
                    .transition(GenericTransitionOptions.with(AnimationFactory.getTransition(AnimationFactory.AnimationType.FADE_IN)))
                    .dontTransform()
                    .listener((ImageRequestListener) status -> Timber.e("%s, glide listener status: %s", TAG, status.toString()))
                    .into(userImage);

            rootView.setOnClickListener(v -> {
                user.setTransitionBundle(getTransitionOptions().toBundle());
                selectedUserLive.postValue(user);
            });
        }
    }
}

File: app/src/main/java/com/icostel/arhitecturesample/ui/listusers/UserAdapter.kt
package com.icostel.arhitecturesample.ui.listusers

import android.app.Activity
import android.app.ActivityOptions
import android.content.Context
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.GenericTransitionOptions
import com.icostel.arhitecturesample.R
import com.icostel.arhitecturesample.di.modules.GlideApp
import com.icostel.arhitecturesample.view.model.User
import com.icostel.commons.utils.AnimationFactory
import com.icostel.commons.utils.bind
import com.icostel.commons.utils.livedata.SingleLiveEvent
import timber.log.Timber
import java.util.*

class UserAdapter(private val context: Context) : RecyclerView.Adapter<UserAdapter.UserViewHolder>() {

    private val users = ArrayList<User>()
    internal var selectedUserLive = SingleLiveEvent<User>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val charityItemView = LayoutInflater.from(this.context).inflate(R.layout.item_user, parent, false)
        return UserViewHolder(charityItemView)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        holder.bindUserViewHolder(users[position])
    }

    fun updateUserList(newUserList: List<User>) {
        Timber.d("updateUserList(), size: %d", newUserList.size)
        val diffCallback = UserDiffCallback(users, newUserList)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        users.clear()
        users.addAll(newUserList)
        notifyDataSetChanged()
        // push only the new items after the diff is done
        diffResult.dispatchUpdatesTo(this)
    }

    override fun getItemCount(): Int {
        return users.size
    }

    inner class UserViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private var rootView: View = itemView.bind(R.id.root_view)
        private var firstName: TextView = itemView.bind(R.id.first_name)
        private var userImage: ImageView = itemView.bind(R.id.user_image)
        private var age: TextView = itemView.bind(R.id.age)

        private val transitionOptions: ActivityOptions
            get() = ActivityOptions.makeSceneTransitionAnimation(itemView.context as Activity, userImage, "user_image_transition")

        fun bindUserViewHolder(user: User) {
            firstName.text = user.firstName
            age.text = context.getString(R.string.age, user.age)

            if (TextUtils.isEmpty(user.resourceUrl)) {
                Timber.d("no user image available")
            } else {
                GlideApp.with(context)
                        .load(user.resourceUrl)
                        .transition(GenericTransitionOptions.with<Drawable>(AnimationFactory.getTransition(AnimationFactory.AnimationType.FADE_IN)))
                        .dontTransform()
                        .into(userImage)
            }

            rootView.setOnClickListener {
                user.transitionBundle = transitionOptions.toBundle()
                selectedUserLive.postValue(user)
            }
        }
    }
}
